File: llm_written_v1/gpt-5/smith-waterman/java/AlgoParallel.java
import java.util.concurrent.*;
import java.util.concurrent.atomic.AtomicReference;

// Parallel Smith-Waterman (local alignment, linear gap)
public class AlgoParallel {
    private final int matchScore;
    private final int mismatchScore;
    private final int gapScore;
    private final int maxWorkers; // cap to available processors
    private final long seqThresholdCells; // tiny-input sequential fallback threshold

    public AlgoParallel(int matchScore, int mismatchScore, int gapScore) {
        this(matchScore, mismatchScore, gapScore,
                Math.max(1, Runtime.getRuntime().availableProcessors()),
                50_000L);
    }

    public AlgoParallel(int matchScore, int mismatchScore, int gapScore, int maxWorkers, long seqThresholdCells) {
        this.matchScore = matchScore;
        this.mismatchScore = mismatchScore;
        this.gapScore = gapScore;
        this.maxWorkers = Math.max(1, maxWorkers);
        this.seqThresholdCells = Math.max(0L, seqThresholdCells);
    }

    // Deterministic anti-diagonal wavefront parallelization with reusable worker threads and barrier
    public int[][] constructMatrix(String query, String reference) {
        final int n = query.length() + 1; // rows
        final int m = reference.length() + 1; // cols
        final int[][] H = new int[n][m];

        // trivial/edge: empty sequences or tiny problems → sequential
        long cells = (long) (n - 1) * (m - 1);
        if (cells <= seqThresholdCells || maxWorkers == 1) {
            return constructMatrixSequential(query, reference);
        }

        final char[] A = query.toCharArray();
        final char[] B = reference.toCharArray();

        final int usableWorkers = Math.max(1, Math.min(maxWorkers, Runtime.getRuntime().availableProcessors()));
        final ExecutorService pool = Executors.newFixedThreadPool(usableWorkers);
        final CyclicBarrier barrier = new CyclicBarrier(usableWorkers);
        final AtomicReference<Throwable> errorRef = new AtomicReference<>(null);
        final int maxD = (n - 1) + (m - 1);

        for (int wid = 0; wid < usableWorkers; wid++) {
            final int workerId = wid;
            pool.execute(() -> {
                try {
                    for (int d = 2; d <= maxD; d++) {
                        final int iStart = Math.max(1, d - (m - 1));
                        final int iEnd = Math.min(n - 1, d - 1);
                        final int diagLen = Math.max(0, iEnd - iStart + 1);
                        // stripe this diagonal by workerId
                        for (int k = workerId; k < diagLen; k += usableWorkers) {
                            int i = iStart + k;
                            int j = d - i;
                            int scoreDiagonal = H[i - 1][j - 1] + (A[i - 1] == B[j - 1] ? matchScore : mismatchScore);
                            int scoreUp = H[i - 1][j] + gapScore;
                            int scoreLeft = H[i][j - 1] + gapScore;
                            int v = scoreDiagonal;
                            if (scoreUp > v) v = scoreUp;
                            if (scoreLeft > v) v = scoreLeft;
                            H[i][j] = Math.max(0, v);
                        }
                        barrier.await(); // end of diagonal
                    }
                } catch (Throwable t) {
                    errorRef.compareAndSet(null, t);
                    // try to trip barrier so others can exit
                    try { barrier.await(1, TimeUnit.MILLISECONDS); } catch (Exception ignored) {}
                }
            });
        }

        pool.shutdown();
        try {
            boolean done = pool.awaitTermination(5, TimeUnit.MINUTES);
            if (!done) throw new RuntimeException("Parallel SW timeout");
        } catch (InterruptedException e) {
            Thread.currentThread().interrupt();
            throw new RuntimeException("Parallel SW interrupted", e);
        }

        Throwable err = errorRef.get();
        if (err != null) {
            if (err instanceof RuntimeException) throw (RuntimeException) err;
            throw new RuntimeException(err);
        }

        return H;
    }

    // Sequential fallback used for very small problems
    private int[][] constructMatrixSequential(String query, String reference) {
        int n = query.length() + 1;
        int m = reference.length() + 1;
        int[][] H = new int[n][m];
        char[] A = query.toCharArray();
        char[] B = reference.toCharArray();
        for (int i = 1; i < n; i++) {
            for (int j = 1; j < m; j++) {
                int scoreDiagonal = H[i - 1][j - 1] + (A[i - 1] == B[j - 1] ? matchScore : mismatchScore);
                int scoreUp = H[i - 1][j] + gapScore;
                int scoreLeft = H[i][j - 1] + gapScore;
                int v = scoreDiagonal;
                if (scoreUp > v) v = scoreUp;
                if (scoreLeft > v) v = scoreLeft;
                H[i][j] = Math.max(0, v);
            }
        }
        return H;
    }

    public int[] findHighestScore(int[][] H) {
        int maxScore = 0;
        int maxI = 0;
        int maxJ = 0;
        for (int i = 0; i < H.length; i++) {
            for (int j = 0; j < H[i].length; j++) {
                if (H[i][j] > maxScore) {
                    maxScore = H[i][j];
                    maxI = i;
                    maxJ = j;
                }
            }
        }
        return new int[]{maxI, maxJ};
    }

    public AlignmentResult traceback(int[][] H, String query, String reference) {
        StringBuilder alignedA = new StringBuilder();
        StringBuilder alignedB = new StringBuilder();

        int[] maxPos = findHighestScore(H);
        int i = maxPos[0];
        int j = maxPos[1];
        int score = H[i][j];

        int totalMatch = 0;
        int totalAlignment = 0;

        while (i > 0 && j > 0) {
            int currentScore = H[i][j];
            if (currentScore == 0) break;

            int diagonalScore = H[i - 1][j - 1];
            int upScore = H[i - 1][j];
            int leftScore = H[i][j - 1];

            int expectedDiagonal = diagonalScore + (query.charAt(i - 1) == reference.charAt(j - 1) ? matchScore : mismatchScore);

            if (currentScore == expectedDiagonal) {
                alignedA.append(query.charAt(i - 1));
                alignedB.append(reference.charAt(j - 1));
                totalAlignment++;
                if (query.charAt(i - 1) == reference.charAt(j - 1)) totalMatch++;
                i--; j--;
            } else if (currentScore == upScore + gapScore) {
                alignedA.append(query.charAt(i - 1));
                alignedB.append('-');
                totalAlignment++;
                i--;
            } else if (currentScore == leftScore + gapScore) {
                alignedA.append('-');
                alignedB.append(reference.charAt(j - 1));
                totalAlignment++;
                j--;
            } else {
                break;
            }
        }

        alignedA.reverse();
        alignedB.reverse();

        double percentageIdentity = (totalAlignment > 0) ? ((double) totalMatch / totalAlignment) * 100.0 : 0.0;

        return new AlignmentResult(alignedA.toString(), alignedB.toString(), score, percentageIdentity);
    }

    public AlignmentResult findAlignment(String query, String reference) {
        int[][] H = constructMatrix(query, reference);
        return traceback(H, query, reference);
    }

    // Same result container as sequential version
    public static class AlignmentResult {
        public final String alignedA;
        public final String alignedB;
        public final int score;
        public final double identity;
        public AlignmentResult(String alignedA, String alignedB, int score, double identity) {
            this.alignedA = alignedA;
            this.alignedB = alignedB;
            this.score = score;
            this.identity = identity;
        }
    }
}

File: llm_written_v1/gpt-5/smith-waterman/rust/algo_parallel.rs
// Parallel Smith-Waterman (deterministic), using anti-diagonal wavefront and std::thread scoped workers
// Public API matches the sequential version. For small inputs or single-core, falls back to sequential.

use std::cmp::max;
use std::thread;

pub struct SmithWaterman {
    match_score: i32,
    mismatch_score: i32,
    gap_score: i32,
}

#[derive(Debug, Clone)]
pub struct AlignmentResult {
    pub aligned_a: String,
    pub aligned_b: String,
    pub score: i32,
    pub identity: f64,
}

impl SmithWaterman {
    pub fn new(match_score: i32, mismatch_score: i32, gap_score: i32) -> Self {
        SmithWaterman {
            match_score,
            mismatch_score,
            gap_score,
        }
    }

    fn construct_matrix_sequential(&self, query: &str, reference: &str) -> Vec<Vec<i32>> {
        let n = query.len() + 1;
        let m = reference.len() + 1;

        let mut h = vec![vec![0; m]; n];

        if n == 0 || m == 0 {
            return h;
        }

        let query_chars: Vec<char> = query.chars().collect();
        let reference_chars: Vec<char> = reference.chars().collect();

        for i in 1..n {
            for j in 1..m {
                let score_diagonal = h[i - 1][j - 1]
                    + if query_chars[i - 1] == reference_chars[j - 1] {
                        self.match_score
                    } else {
                        self.mismatch_score
                    };

                let score_up = h[i - 1][j] + self.gap_score;
                let score_left = h[i][j - 1] + self.gap_score;

                h[i][j] = 0.max(score_diagonal.max(score_up.max(score_left)));
            }
        }

        h
    }

    pub fn construct_matrix(&self, query: &str, reference: &str) -> Vec<Vec<i32>> {
        let n = query.len() + 1;
        let m = reference.len() + 1;

        let total = n.saturating_mul(m);
        let threads = std::thread::available_parallelism().map(|nz| nz.get()).unwrap_or(4);
        let min_work_for_parallel: usize = 1_000_000; // raise threshold to avoid overhead on moderate sizes

        if threads <= 1 || total < min_work_for_parallel || n <= 1 || m <= 1 {
            return self.construct_matrix_sequential(query, reference);
        }

        let query_chars: Vec<char> = query.chars().collect();
        let reference_chars: Vec<char> = reference.chars().collect();

        // Matrix initialized with zeros
        let mut h = vec![vec![0_i32; m]; n];

        // Process anti-diagonals k = 2..=(n-1 + m-1)
        let first_k = 2_usize;
        let last_k = (n - 1) + (m - 1);

        for k in first_k..=last_k {
            // Determine range of i for this diagonal
            let i_start = 1.max(k.saturating_sub(m - 1));
            let i_end = (n - 1).min(k - 1);
            if i_start > i_end {
                continue;
            }

            // Prepare indices in deterministic order (increasing i)
            let mut diag_indices: Vec<(usize, usize)> = Vec::with_capacity(i_end - i_start + 1);
            for i in i_start..=i_end {
                let j = k - i;
                if j >= 1 && j <= m - 1 {
                    diag_indices.push((i, j));
                }
            }
            if diag_indices.is_empty() {
                continue;
            }

            let work_len = diag_indices.len();
            let used_threads = threads.min(work_len);
            let chunk_size = (work_len + used_threads - 1) / used_threads;

            let mut results: Vec<(usize, i32)> = Vec::with_capacity(work_len);

            // Scoped threads so we can borrow h/query/reference immutably
            thread::scope(|s| {
                let h_ref: &Vec<Vec<i32>> = &h; // read-only during this diagonal
                let q_ref: &Vec<char> = &query_chars;
                let r_ref: &Vec<char> = &reference_chars;
                let mut handles = Vec::with_capacity(used_threads);
                for t in 0..used_threads {
                    let start = t * chunk_size;
                    if start >= work_len {
                        break;
                    }
                    let end = ((t + 1) * chunk_size).min(work_len);
                    let slice = &diag_indices[start..end];
                    let handle = s.spawn(move || -> Vec<(usize, i32)> {
                        let mut local: Vec<(usize, i32)> = Vec::with_capacity(slice.len());
                        for (offset, &(i, j)) in slice.iter().enumerate() {
                            let pos = start + offset; // position within diagonal
                            let diagonal_score = h_ref[i - 1][j - 1]
                                + if q_ref[i - 1] == r_ref[j - 1] {
                                    self.match_score
                                } else {
                                    self.mismatch_score
                                };
                            let up_score = h_ref[i - 1][j] + self.gap_score;
                            let left_score = h_ref[i][j - 1] + self.gap_score;
                            let val = 0.max(max(diagonal_score, max(up_score, left_score)));
                            local.push((pos, val));
                        }
                        local
                    });
                    handles.push(handle);
                }
                for handle in handles {
                    let mut part = handle.join().expect("thread panicked");
                    results.append(&mut part);
                }
            });

            // Place results in deterministic order (by position within diagonal)
            results.sort_by_key(|&(pos, _)| pos);
            for (idx_in_diag, &(_, val)) in results.iter().enumerate() {
                let (i, j) = diag_indices[idx_in_diag];
                h[i][j] = val;
            }
        }

        h
    }

    pub fn find_highest_score(&self, h: &[Vec<i32>]) -> (usize, usize) {
        let mut max_score = 0;
        let mut max_pos = (0, 0);

        for (i, row) in h.iter().enumerate() {
            for (j, &val) in row.iter().enumerate() {
                if val > max_score {
                    max_score = val;
                    max_pos = (i, j);
                }
            }
        }

        max_pos
    }

    pub fn traceback(&self, h: &[Vec<i32>], query: &str, reference: &str) -> AlignmentResult {
        let mut aligned_a = String::new();
        let mut aligned_b = String::new();

        let (mut i, mut j) = self.find_highest_score(h);
        let score = h[i][j];

        let mut total_match = 0;
        let mut total_alignment = 0;

        let query_chars: Vec<char> = query.chars().collect();
        let reference_chars: Vec<char> = reference.chars().collect();

        while i > 0 && j > 0 {
            let current_score = h[i][j];

            if current_score == 0 {
                break;
            }

            let diagonal_score = h[i - 1][j - 1];
            let up_score = h[i - 1][j];
            let left_score = h[i][j - 1];

            let expected_diagonal = diagonal_score
                + if query_chars[i - 1] == reference_chars[j - 1] {
                    self.match_score
                } else {
                    self.mismatch_score
                };

            if current_score == expected_diagonal {
                aligned_a.push(query_chars[i - 1]);
                aligned_b.push(reference_chars[j - 1]);
                total_alignment += 1;

                if query_chars[i - 1] == reference_chars[j - 1] {
                    total_match += 1;
                }
                i -= 1;
                j -= 1;
            } else if current_score == up_score + self.gap_score {
                aligned_a.push(query_chars[i - 1]);
                aligned_b.push('-');
                total_alignment += 1;
                i -= 1;
            } else if current_score == left_score + self.gap_score {
                aligned_a.push('-');
                aligned_b.push(reference_chars[j - 1]);
                total_alignment += 1;
                j -= 1;
            } else {
                break;
            }
        }

        aligned_a = aligned_a.chars().rev().collect();
        aligned_b = aligned_b.chars().rev().collect();

        let percentage_identity = if total_alignment > 0 {
            (total_match as f64 / total_alignment as f64) * 100.0
        } else {
            0.0
        };

        AlignmentResult {
            aligned_a,
            aligned_b,
            score,
            identity: percentage_identity,
        }
    }

    pub fn find_alignment(&self, query: &str, reference: &str) -> AlignmentResult {
        let h = self.construct_matrix(query, reference);
        self.traceback(&h, query, reference)
    }
}
